### PR TITLE
Binary FlatGeobuf parsing

### PR DESCRIPTION
**Pull request by kylebarron**
For #716 

- [x] Adds reprojection support (just for GeoJSON output, not binary output at the moment)
- [x] Adds initial docs and direct-to-binary parsing
- [ ] Add test files

### CODE REVIEW

File: modules/flatgeobuf/package.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
   },
   "dependencies": {
     "@loaders.gl/loader-utils": "^2.1.1",
-    "flatgeobuf": "^3.1.0"
+    "flatgeobuf": "^3.5.0"
   },
   "devDependencies": {
     "esm": "^3.2.25"
```
Comment (kylebarron) [inline modules/flatgeobuf/package.json:39]: It should be possible to use uncompiled ES6 files from `flatgeobuf/lib` with `esm`, but I'm not sure how to wire that up.

File: modules/flatgeobuf/src/lib/binary-geometries.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  // Assumes all parts of the multipolygon have the same size
+  const positionSize = parsedParts[0].positions.size;
+
+  for (const parsedPart of parsedParts) {
+    concatPositions.set(parsedPart.positions.value, positionCounter * positionSize);
+
+    // For indices, need to add positionCounter so that position indices are
+    // correct in the concatenated positions
+    concatPrimitivePolygonIndices.set(
+      parsedPart.primitivePolygonIndices.value.subarray(1).map(x => x + positionCounter),
```
Comment (kylebarron) [inline modules/flatgeobuf/src/lib/binary-geometries.js:107]: I get `no-loop-func` eslint errors here https://eslint.org/docs/rules/no-loop-func

File: modules/flatgeobuf/package.json (hunk 1 above)
Comment (ibgreen) [reply modules/flatgeobuf/package.json:39]: isn't `esm` an equivalent of `reify` which we currently use. It enables ES6 imports in Node.js.

I don't think it has any effect in the browser.

Users on older browsers will not be able to use this module if the source is not transpiled. 

> to support geojson and binary parsing we need to take from two overlapping dist bundles.

Why is that ?

File: modules/flatgeobuf/src/lib/binary-geometries.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  if (!z) {
+    return {value: xy, size: 2};
+  }
+
+  if (z.length * 2 !== xy.length) {
+    throw new Error("Z array must be half XY array's length");
+  }
+  const totalLength = xy.length + z.length;
+
+  const xyz = new Float64Array(totalLength);
```
Comment (ibgreen) [inline modules/flatgeobuf/src/lib/binary-geometries.js:138]: Generating a typed array for every little feature is unlikely to gain us any performance advantage.

Is a 2 pass approach out of the question?

File: modules/flatgeobuf/src/lib/binary-properties.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
+/* global TextDecoder */
+
+// Eventually will be re-exported from generic module, hopefully
```
Comment (ibgreen) [inline modules/flatgeobuf/src/lib/binary-properties.js:3]: You mean from the `flatgeobuf` module?

File: modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    coordinates: 'wgs84',
+    tileIndex: {
+      x: 0,
+      y: 0,
+      z: 0
+    }
+  }
+};
+
+const geoJSONfeatures = await load(url, MVTLoader, loaderOptions);
```
Comment (ibgreen) [inline modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md:38]: search replace MVT with case insensitive settings.

File: modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+## Outputs
+
+### GeoJSON
+
+The parser will return an array of [GeoJSON objects](https://tools.ietf.org/html/rfc7946) with WGS84 coordinates and feature properties from MVT if `coordinates` property is set to `wgs84` and `tileIndex` properties are present.
+
+```js
+import {MVTLoader} from '@loaders.gl/mvt';
```
Comment (ibgreen) [inline modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md:48]: ditto

File: modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import {load} from '@loaders.gl/core';
+
+const loaderOptions = {
+  mvt: {
+    coordinates: 'wgs84',
+    tileIndex: {
```
Comment (ibgreen) [inline modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md:54]: Fix

File: modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+const geoJSONfeatures = await load(url, MVTLoader, loaderOptions);
+```
+
+## Options
+
+| Option          | Type   | Default | Description                                                                                                                                                                                                                                                                            |
+| --------------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| mvt.coordinates | String | `local` | When set to `wgs84`, the parser will return a flat array of GeoJSON objects with coordinates in longitude, latitude decoded from the provided tile index. When set to `local`, the parser will return a flat array of GeoJSON objects with local coordinates decoded from tile origin. |
```
Comment (ibgreen) [inline modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md:69]: fix

File: modules/flatgeobuf/src/lib/parse-flatgeobuf.js [hunk 9]
```diff
@@ -1,5 +1,27 @@
-// import {deserialize} from 'flatgeobuf/lib/geojson/featurecollection';
-import {deserialize, deserializeStream} from 'flatgeobuf/dist/flatgeobuf-geojson.min';
+import {
+  deserialize as deserializeGeoJson,
+  deserializeStream as deserializeStreamGeoJson
+} from 'flatgeobuf/dist/flatgeobuf-geojson.min';
+import {
+  deserialize as deserializeGeneric,
+  deserializeStream as deserializeStreamGeneric
+} from 'flatgeobuf/dist/flatgeobuf.min';
```
Comment (kylebarron) [inline modules/flatgeobuf/src/lib/parse-flatgeobuf.js:8]: Re https://github.com/visgl/loaders.gl/pull/884/files#r466768308, 

> > to support geojson and binary parsing we need to take from two overlapping dist bundles.

> Why is that ?

There are three `dist` bundles exported from `flatgeobuf`: 

- `flatgeobuf/dist/flatgeobuf-ol.min.js`: parses Flatgeobuf to OpenLayers objects
- `flatgeobuf/dist/flatgeobuf-geojson.min.js`: parses Flatgeobuf to GeoJSON objects
- `flatgeobuf/dist/flatgeobuf.min.js`: the _generic_ distribution. You pass in a function that manually parses the `flatbuffers` objects

If you want simple two-line parsing to geojson, you use the `flatgeobuf-geojson.min.js` file. If you want to parse directly to binary, you need the `flatgeobuf.min.js` file where you can pass your own parsing function.

The ways around this are either: 

- only use `flatgeobuf.min.js` and copy all the geojson parsing code from the flatgeobuf repo
- Import from the uncompiled ES6 `flatgeobuf/lib/*` and let our build tool do the tree shaking

The latter would be preferable, but I was having issues loading from `flatgeobuf/lib/*` in Node.

File: modules/flatgeobuf/src/lib/binary-properties.js (hunk 4 above)
Comment (kylebarron) [reply modules/flatgeobuf/src/lib/binary-properties.js:3]: See [below](https://github.com/visgl/loaders.gl/pull/884/files#r466772726). If I get `esm`/`reify` working I can import this file from `flatgeobuf/lib/` instead.

File: modules/flatgeobuf/src/lib/binary-geometries.js (hunk 3 above)
Comment (kylebarron) [reply modules/flatgeobuf/src/lib/binary-geometries.js:138]: The Flatgeobuf module parses features individually. For non-streaming you could theoretically call `deserialize` twice, once to collect lengths, again to set positions within an instantiated typed array.

I'm not sure how you could do that with streaming. Flatgeobuf's `deserializeStream` abstracts this from the user. The parsing function you pass operates on a single feature at a time.

Note that for `positions` of all 2D geometry types except for MultiPolygon, no copy should be needed. They're already transported in a flat 2D array.

File: modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md (hunk 5 above)
Comment (kylebarron) [reply modules/flatgeobuf/docs/api-reference/flatgeobuf-loader.md:38]: Yes, as you've noticed I put docs off till the rest of the PR is ready.

**Comment by kylebarron** [comment]
> Is a 2 pass approach out of the question?

I thought about this a little more. You could first "collect" underlying flatbuffers with something like:

```js
const flatbufferFeatures = [];
deserialize(arrayBuffer, (feature, header) => {
  flatbufferFeatures.push(feature)
})
```

Then it would be simple to do two passes on the `flatbufferFeatures` array. For streaming you could collect these into batches of `n` features at a time, then run `flatgeobufBatchToBinary` on them.

**Comment by ibgreen** [comment]
Close in favor of rebased #1942

**Pull request closed without merging**